### PR TITLE
fix(applier): Modify apply layer to merge custom resources as well

### PR DESCRIPTION
**Pull request by ankk13**
With addition for analyzer group, a CustomResource identifier is added to support any custom analyzer.
This commit allows merging of custom resources

### CODE REVIEW

**Review by knqyf263 — CHANGES_REQUESTED** [review]
Could you add a test case having CustomResource? Also, the file path should conflict with OS packages or language-specific packages so that we can confirm duplicate file paths work properly.

**Comment by ankk13** [comment]
> Could you add a test case having CustomResource? Also, the file path should conflict with OS packages or language-specific packages so that we can confirm duplicate file paths work properly.

I have updated PR with the required unit tests.

**Comment by ankk13** [comment]
> Could you add a test case having CustomResource? Also, the file path should conflict with OS packages or language-specific packages so that we can confirm duplicate file paths work properly.

I have updated PR with the required unit tests.

**Review by knqyf263 — CHANGES_REQUESTED** [review]

File: applier/applier_test.go [hunk 1]
```diff
@@ -500,6 +504,219 @@ func TestApplier_ApplyLayers(t *testing.T) {
...
+										data: "Common Application type-A php-app/composer.lock",
+									},
+								},
+								{
+									Type:     "type-B",
+									FilePath: "php-app/composer.lock",
+									Data: dummyData{
+										data: "Common Application type-B php-app/composer.lock",
+									},
+								},
```
Comment (knqyf263) [inline applier/applier_test.go:611]: We should check if replacing a previous layer works or not. When the type and file path are the same, the upper layer must replace CustomResources in the previous layer. In this case, the previous layer has `"var/lib/dpkg/status.d/tzdata"` with `type-B`. We should have the same pair with different data.

**Review by knqyf263 — CHANGES_REQUESTED** [review]

File: applier/applier_test.go [hunk 2]
```diff
@@ -500,6 +506,246 @@ func TestApplier_ApplyLayers(t *testing.T) {
...
+									Libraries: []types.Package{
+										{
+											Name:    "guzzlehttp/guzzle",
+											Version: "6.2.0",
+										},
+										{
+											Name:    "symfony/process",
+											Version: "v4.2.7",
+										},
+									},
```
Comment (knqyf263) [inline applier/applier_test.go:603]: Why do we need this object?

File: applier/applier_test.go [hunk 3]
```diff
@@ -500,6 +506,246 @@ func TestApplier_ApplyLayers(t *testing.T) {
...
+									Data: dummyData{
+										data: "Common Package type-A var/lib/dpkg/status.d/tzdata",
+									},
+								},
+								{
+									Type:     "type-B",
+									FilePath: "var/lib/dpkg/status.d/tzdata",
+									Data: dummyData{
+										data: "Common Package type-B var/lib/dpkg/status.d/tzdata",
+									},
```
Comment (knqyf263) [inline applier/applier_test.go:554]: What is this for? We have `type-C` and it seems to be enough.

**Review by knqyf263 — CHANGES_REQUESTED** [review]

File: applier/applier_test.go (hunk 3 above)
Comment (ankk13) [reply applier/applier_test.go:554]: This case was added for case, if same resource on same layer is added by different analyzers, they should be equivalently merged. Tesecase check for intralayer merging. 
Type-C case is specific to same resource, on **different** layer by different analyzer.

Comment (knqyf263) [reply applier/applier_test.go:554]: >if same resource on same layer is added by different analyzers

Type-C seems to satisfy the above condition.

We want to test the following cases.

1. The same resource on the same layer by different analyzers
2. The same resource on the different layers by the same analyzer

The pair of Type-A and Type-C seems to cover both. Or, am I missing something? Can Type-B test something new?

Comment (ankk13) [reply applier/applier_test.go:554]: Above mentioned testcases are satisfied by Type-A and Type-C. 
But following testcase is missed. 
Consider a single layer with same resource(1 as Type-A and 1 as Type-C) but detected by different analyzer. If we have bug introduced in it merging this data. The above testcase may mask this failure, as the data will be overridden by merge on the next layer.

Comment (ankk13) [reply applier/applier_test.go:554]: Above mentioned testcases are satisfied by Type-A and Type-C. 
But following testcase is missed. 
Consider a single layer with same resource(1 as Type-A and 1 as Type-C) but detected by different analyzer. If we have bug introduced in it merging this data(intralayer). The above testcase may mask this failure, as the data will be overridden by merge on the next layer.

Comment (knqyf263) [reply applier/applier_test.go:554]: What if we move Type-A to the upper layer?

- upper layer
  - foo.txt
    - Type-A
    - Type-C
- lower layer
  - foo.txt
    - Type-C

Comment (ankk13) [reply applier/applier_test.go:554]: okey. I will update it accordingly

**Review by knqyf263 — CHANGES_REQUESTED** [review]

File: applier/applier_test.go [hunk 4]
```diff
@@ -500,6 +506,174 @@ func TestApplier_ApplyLayers(t *testing.T) {
...
+							CustomResources: []types.CustomResource{
+								{
+									Type:     "type-A",
+									FilePath: "var/lib/dpkg/status.d/tzdata",
+									Data: dummyData{
+										data: "Common Package type-A var/lib/dpkg/status.d/tzdata",
+									},
+								},
+								{
+									Type:     "type-C",
```
[the remark this answers is not in the document]
Comment (ankk13) [reply applier/applier_test.go]: Done

File: applier/applier_test.go [hunk 5]
```diff
@@ -525,6 +699,16 @@ func TestApplier_ApplyLayers(t *testing.T) {
...
 				})
 			}
+
+			sort.Slice(got.CustomResources, func(i, j int) bool {
+				if got.CustomResources[i].FilePath < got.CustomResources[j].FilePath {
+					return true
+				} else if got.CustomResources[i].FilePath == got.CustomResources[j].FilePath {
+					return got.CustomResources[i].Type < got.CustomResources[j].Type
+				}
+				return false
```
Comment (knqyf263) [inline applier/applier_test.go:709]: How about this?
```suggestion
				if  got.CustomResources[i].FilePath == got.CustomResources[j].FilePath {
					return got.CustomResources[i].Type < got.CustomResources[j].Type
				}
				return got.CustomResources[i].FilePath < got.CustomResources[j].FilePath
```

Comment (ankk13) [reply applier/applier_test.go]: yes, more readable. Updated

**Review by knqyf263 — APPROVED** [review]
:shipit:

**Pull request merged**